### PR TITLE
Avoid migration on stateless firewalls

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | Related to #27395
| License       | MIT
| Doc PR        | symfony/symfony-docs#9860

This is a proof-of-concept. Once we agree / are happy, I need to add this to all of the other authentication mechanisms that recently got the session migration code & add tests.

Basically, this avoids migrating the session if the firewall is stateless. There were 2 options to do this:

A) Make the `SessionAuthenticationStrategy` aware of all stateless firewalls. **This is the current approach**
or
B) Make each individual authentication listener aware whether or not *its* firewall is stateless.

### CODE REVIEW

**Comment by stof** [comment]
third option: inject a no-op session strategy in listeners belonging to stateless firewalls

**Comment by stof** [comment]
your first approach requires every listener to set the firewall name as a request attribute before calling the session strategy (which is very easy to miss as it is not part of the explicit API of the session strategy). So IMO, that's not a good approach

**Comment by aschempp** [comment]
I like the noop-idea, It does not even need to be a new no-op class. A second session migration strategy for stateless firewalls would solve that, which gets a `NONE` strategy set in the constructor.

The respective listeners must then get the "stateless migration strategy" injected (instead of https://github.com/weaverryan/symfony/blob/4d5bc448ee2c192c9f4ef964e3006a63a707029e/src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml#L178)

**Comment by weaverryan** [comment]
No op is a great idea - I’ll update

**Comment by weaverryan** [comment]
I've just updated to use a noop approach. I think this is solid. Unless anyone sees a major issue, I'll extend this to all of the other listeners & add some tests.

For `SimplePreAuthenticationListener`, I'm using setter injection because there are already so many optional constructor args. I'm not sure if, in the future, this should or should not be a required dependency. We could say "hey, you need to pass this in if you are storing the token in the session" OR we could eventually force (via some deprecations in 4.2) the user to pass this in, telling them to pass in a noop if they don't want session migration.

**Comment by aschempp** [comment]
Why do we need a noop strategy at all if you can simply not inject a strategy for the same behavior?

**Comment by weaverryan** [comment]
Good question - it’s just a technical work around. The listener service is created in the factory, but the factory doesn’t know whether or not it will live in a stateless firewall. So, it can’t be smart enough (unless I’m missing an idea - very possible) to know if it should or should not inject the session strategy. To work around that, we simply *always* inject a session strategy, but then in SecurityExtension, we change that session strategy to be a noop for stateless firewalls. It’s a tricky dance.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 1]
```diff
@@ -89,6 +89,8 @@ public function load(array $configs, ContainerBuilder $container)
         $this->createAuthorization($config, $container);
         $this->createRoleHierarchy($config, $container);
 
+        $container->setParameter('security.authentication.stateless_firewalls', $this->statelessFirewalls);
```
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: not needed anymore now :)

**Comment by weaverryan** [comment]
This is ready for review. I was very careful, but as I don't believe there are tests on the "factory" classes that build the auth listeners, I appreciate close review.

I added a test for `GuardAuthenticationHandler`. We could also add tests for all the auth listeners - I'm happy to do that, but it would be quite a bit of test refactoring or duplication (as you need to build a test that successfully authenticates, which is often quite a lot of code). But if we feel better with those tests, let me know!

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php [hunk 2]
```diff
@@ -129,15 +130,22 @@ public function handleAuthenticationFailure(AuthenticationException $authenticat
         ));
     }
 
-    private function migrateSession(Request $request)
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @param SessionAuthenticationStrategyInterface $sessionStrategy
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: can be removed

File: src/Symfony/Component/Security/Http/Firewall/AbstractPreAuthenticatedListener.php [hunk 3]
```diff
@@ -96,6 +99,16 @@ public function __construct(TokenStorageInterface $tokenStorage, AuthenticationM
         }
     }
 
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @param SessionAuthenticationStrategyInterface $sessionStrategy
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/AbstractPreAuthenticatedListener.php]: no need for this one :)

File: src/Symfony/Component/Security/Http/Firewall/BasicAuthenticationListener.php [hunk 4]
```diff
@@ -93,15 +96,22 @@ public function handle(GetResponseEvent $event)
         }
     }
 
-    private function migrateSession(Request $request)
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @param SessionAuthenticationStrategyInterface $sessionStrategy
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/BasicAuthenticationListener.php]: to be removed

File: src/Symfony/Component/Security/Http/Firewall/DigestAuthenticationListener.php [hunk 5]
```diff
@@ -117,9 +120,20 @@ public function handle(GetResponseEvent $event)
...
+        $this->migrateSession($request, $token);
 
-        $this->tokenStorage->setToken(new UsernamePasswordToken($user, $user->getPassword(), $this->providerKey));
+        $this->tokenStorage->setToken($token);
+    }
+
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @param SessionAuthenticationStrategyInterface $sessionStrategy
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/DigestAuthenticationListener.php]: boo

File: src/Symfony/Component/Security/Http/Firewall/SimplePreAuthenticationListener.php [hunk 6]
```diff
@@ -62,6 +65,16 @@ public function __construct(TokenStorageInterface $tokenStorage, AuthenticationM
         $this->dispatcher = $dispatcher;
     }
 
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @param SessionAuthenticationStrategyInterface $sessionStrategy
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/SimplePreAuthenticationListener.php]: :P

**Review by nicolas-grekas — COMMENTED** [review]
there is a failure on deps=low, a composer.json needs a tweak somewhere
cheers :)

**Comment by weaverryan** [comment]
Ok, CI is happy!

To summarize: the only downside to this PR is on a technical level: the fact that I'm using setter injection everywhere to avoid trying to add this dependency to the constructor (when every class already has optional deps). I've made the setters final, so, if we want to change this in the future, we could decide to do that with less trouble.

**Comment by nicolas-grekas** [comment]
LGTM. I'd just suggest to move the `final` note to an annotation, as real `final` tokens legit usages (e.g. lazy proxies) and we try to avoid them here usually.

**Comment by weaverryan** [comment]
Done!

I'll make a separate PR for 3.4 for `UsernamePasswordJsonAuthenticationListener`, but it won't block this (that is on my list).

File: src/Symfony/Bundle/SecurityBundle/composer.json [hunk 7]
```diff
@@ -18,7 +18,7 @@
     "require": {
         "php": ">=5.3.9",
         "ext-xml": "*",
-        "symfony/security": "^2.8.41|^3.4.11",
+        "symfony/security": "^2.8.42|^3.4.12",
```
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/composer.json]: I bumped both versions - pretty sure that's correct :)

**Review by chalasr — APPROVED** [review]
I would just make the `setSessionAuthenticationStrategy()` internal to not introduce any new public api in a patch and just fix our own listeners until this reaches master. 👍 anyways

**Comment by weaverryan** [comment]
> I would just make the setSessionAuthenticationStrategy() internal to not introduce any new public api in a patch and just fix our own listeners until this reaches master. 👍 anyways

We could do this, but technically speaking, if anyone uses the components, then they would need to use these methods to migrate the session. So, I'm not sure that we can

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php [hunk 8]
```diff
@@ -129,15 +130,22 @@ public function handleAuthenticationFailure(AuthenticationException $authenticat
         ));
     }
 
-    private function migrateSession(Request $request)
+    /**
+     * Call this method if your authentication token is stored to a session.
+     *
+     * @final since version 2.8
```
Comment (fabpot) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: "since version 2.8" should be removed. I don't think we are doing that anywhere else.

**Review by fabpot — APPROVED** [review]
with minor comment

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php (hunk 8 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: We do when we want to hint that the `@final` has been added after the method has been introduced. i.e. when it's deprecated to extend from it.
Which is *not* the case here, so yes, this has to be removed.

**Comment by fabpot** [comment]
Thank you @weaverryan.

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php (hunk 8 above)
Comment (fabpot) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: see 5c2b2bb2ce
